Various parse corrections

claytonrcarter opened the pull request:
Hi! I've encountered a few parsing gotchas while working on an in-development language server (and accompanying Zed extension) for Ledger. This PR includes fixes and added support for everything I've run into so far. Each commit should stand on its own and you should be able so review each one individually. (That's probably the easiest way to grok this.)

 If you'd prefer to see these all submitted separately, I'd be happy to break this apart into separate PRs with fewer changes in each. Just let me know. 

**Changes**
- fix: support accounts that contain single letter "words" (eg `Food & Lodging`)
- fix: empty notes (though useless) were not being recognized as notes (eg `Foo  $1.00 ;\n`)
- fix: notes with no leading whitespace were not being recognized as notes (eg `;foo` vs `; foo`)
- fix: notes that follow a payee name were not being recognized as notes
  - eg `Payee ; foo` => `(payee)` node would contain all of `Payee ; foo`
- fix: lines containing only whitespace that follow a transaction were being incorrectly parsed as an actual posting
  - this caused adjacent tranasctions to "run together" and be parsed as a single transaction
- fix: multiple whitespaces before an ×act code was causing the code to be skipped and instead included as part of the payee name
  - eg `2024-11-18  (code) payee` was being parsed such that the `(payee)` node contained `(code) payee`
- fix: payee names can contain `!` and `*`
- ~~feat: add support for lot prices~~ **edit** I moved this to #34
  - eg https://ledger-cli.org/doc/ledger3.html#Fixing-Lot-Prices

Some of these changes don't come up often in "final" documents, but they were causing funky highlighting during editing. Eg notes we're colored correctly when first inserting `;`, but would be fixed after adding a space. Or the issue w/ "blank lines" would cause transactions to highlight incorrectly, but then that would "fix" after saving and having trailing whitespace removed. That's why some of these may seem like "why would someone ever do that?!" 😄 

~~The only gotcha that I'm aware of is that the last commit, to add lot price support, *is* correctly parsing lot prices, but it's also inserting an error for the space preceding the lot price, even though I copied that nearly verbatim from the `price` node.~~

claytonrcarter committed:
fix: make note parsing more flexible

Although a note w/o any content isn't especially useful, it is still
valid according to Ledger. Moreover, supporting such notes fixes glitchy
syntax highlighting when comments are being added: without this fix,
adding a note to a transaction or posting would cause the next posting
to be incorrectly parsed as a comment.

The change for "no leading whitespace" is basically the same, except
that's just stylistic: ledger doesn't care if a note has leading
whitespace, so perfectly valid notes can cause parse errors, which leads
to the same issue as above: the next posting is incorrectly highlighted
as a comment.

claytonrcarter committed:
fix: allow special chars in payees

Bank statements often include asterisks in the transaction name, and Ledger
seems to place few limits on what characters can make up a payee name, but this
grammar had been excluding ! and *. This removes that restriction so that
payees can now contain ! and *.

Payees still cannot *start* with !, * or (, presumably to disambiguate with
codes and statuses. Interestingly, Ledger doesn't seem to share this
limitation, and a payee can start start with a !, * or even (, so long as
that payee is preceeded by a code. For example, Ledger doesn't seem to mind
any of these:

2024-11-18 () ( Payee)
2024-11-18 () ! Payee
2024-11-18 () * Payee

"Interestingly" because these are obviously problematic names (removing the
empty code would cause them all to mean something else), not because I think we
should consider supporting them. :)

claytonrcarter committed:
fix: notes following payee

Notes that follow the payee are incorrectly being consumed as part of the
payee. There is a test that includes such a note, but the error was missed
because the assertion only included 1 note node, not 2 as there should have
been all along.

This change correctly parses these notes separate from the payee, includes a
test to make to make future regressions more obvious, and corrects the existing
test to expect 2 note nodes.

claytonrcarter committed:
fix: don't consume whitespace-only lines as postings

Not a true fix (note that the test includes an error), but this at least gets
the parser to recognize 2 transactions w/ 1 posting each instead of 1
transaction w/ 3 postings.

claytonrcarter committed:
fix: improve handling of whitespace before codes

When multiple spaces were included before the xact code, the code was being
consumed and included as part of the payee.

This fixes the issue by clarifying that payees cannot start with a whitespace
char, and adds a test for this case.

cbarrete commented on test/corpus/misc.txt:
```diff
@@ -136,7 +136,7 @@ effective date
 ================================================================================
 
 2020-12-02=2020-12-03
-    Expenses:Food:Groceries    ; [=2020-12-04]
```
Is this a mistake/temporary local edit that you forgot to revert?

claytonrcarter replied:
No, not a mistake. I moved that to the `comments_and_notes.txt` tests (title: "with effective date") because it seemed to fit in there. I recall that I broke it while working on my changes to notes, but I was only running the `comments_and_notes` tests so didn't notice until after. But now I'm questioning my judgement on moving it. It's certainly not important (or related to this change!) and I can revert it if you'd prefer.

cbarrete replied:
No it's fine, I had just missed that it was still tested elsewhere. All good!

cbarrete commented on test/corpus/accounts.txt:
```diff
@@ -65,3 +65,43 @@ single letter account name
+================================================================================
+
+2021-01-01
+    Un et Duex
```
Deux* :p (feel free to leave as is, I don't want to waste your time with this and the typo is funny to me)

claytonrcarter replied:
Je suis désolé! I just pushed a fix for this. 🤦 🇫🇷

cbarrete commented:
Thanks for the very high quality contribution (you even fixed something that should have been caught by an existing-yet-buggy test...)!
Everything looks fine to me, except for the removal of the test that I pointed out.
For context, my initial implementation was based on a grammar file I found in the upstream ledger repo, only to find out later that it was quite outdated. Regardless, it makes a lot of sense for this grammar to be more tolerant than the upstream one I think. I'd reconsider if someone wanted to build strict tooling with `tree-sitter-ledger`, but then again, they should probably use the upstream lib or Python bindings.

claytonrcarter committed:
feat: accounts with single letter words

Single letter account names seem to work fine, as do accounts with spaces, but
accounts with single letter words surrounded by spaces do not work without this
change.

claytonrcarter commented:
> Thanks for the very high quality contribution 

And thank you for the prompt review!

I've fixed the typo, and I'll gladly un-move that effective date test if you'd prefer. 

Thank you!
